### ISSUE TITLE
Config unavailable on startup in PC facade

### ISSUE THREAD

**Issue by msteiger** [MovingBlocks/Terasology#1843]
The game crashes with a NPE when started with the server port command line argument. This is due to
```java
engine.getFromEngineContext(Config.class);
```
returning `null` (Terasology.java - line 132).

Sending a ping to @flo hoping that he knows a quick fix for it.

**Comment by flo** [comment]
I had a look at it. The issue is that since #1835 the configuration object gets created in a subsystem. The configuration object gets however used earlier when you pass certain parameters to terasology.

One possibly solution might be switching from command line argumetns to java system properties. Or the configuration sub system needs to be created explictly so that we can pass some arguments to it.

@immortius  What do you think about this?

**Comment by immortius** [comment]
I'll have a look/think. Perhaps this is also an opportunity to look at what scala's config support offers.

**Comment by immortius** [comment]
Ok, I'm inclined to use a java system properties, and switch to using https://github.com/typesafehub/config which supports automatically overriding config properties with java system properties.

**Comment by immortius** [comment]
Hmm, not working so well due to a lack of support of saving configuration from beans.

I guess stepping back and looking at use cases:

1. Want to load config from a user file if available.
2. This should be applied on top of default configuration, which is also used if the user file isn't available. Any config values that are missing come from the default config.
3. It should be possible to override at least some values via commandline, for the current session only. These should not be saved.
4. Configuration need to be editable, and the edits should... override the overrides? Or not? Not would be easier, and more consistent in behavior.

So this could be handled by....

1. Loading the file into a ConfigurationDocument, which can be edited and saved.
2. Applying the ConfigurationDocument over the reference config and under the overrides, to produce the actual configuration to use. This would need to be redone whenever the configuration is altered.
3. The final configuration could be loaded directly, or loaded into beans (as we have been doing).

**Comment by immortius** [comment]
Yeah, I'm going to give up on typesafehub's config. Too many issues, both in a mismatch on what it is aimed for (a configuration that isn't editable in the app itself) and due to some odd omissions in what it supports (doesn't support floats for instance).

**Comment by immortius** [comment]
TODO:

 - [ ] Extract default config into sample file
 - [ ] Load up default config into json
 - [ ] Apply actual file, if available, onto default config. Use this for the editable, saveable config
 - [ ] Apply overrides from java properties on top of the above config. This forms the config that should be used.
 - [ ] When config is saved, update used config.
 - [ ] When module environment changes, apply default configuration from the used modules to introduce module specific configuration, updating used config.
 - [ ] Consider improvements to management of module-specific config.
 - [ ] Consider improvements to prevent loss of module-specific config when switching using older versions.

**Comment by Cervator** [comment]
@immortius do you consider this still outstanding or entirely addressed by #1844 ? If the remaining items are still valid would they be better served as future todos rather than being left here?
